crossplane/function/resource.py

```diff
@@ -38,7 +38,15 @@ def update(r: fnv1.Resource, source: dict | structpb.Struct | pydantic.BaseModel
     """
     match source:
         case pydantic.BaseModel():
-            r.resource.update(source.model_dump(exclude_defaults=True, warnings=False))
+            data = source.model_dump(exclude_defaults=True, warnings=False)
+            # In Pydantic, exclude_defaults=True in model_dump excludes fields
+            # that have their value equal to the default. If a field like
+            # apiVersion is set to its default value 's3.aws.upbound.io/v1beta2'
+            # (and not explicitly provided during initialization), it will be
+            # excluded from the serialized output.
+            data['apiVersion'] = source.apiVersion
+            data['kind'] = source.kind
+            r.resource.update(data)
```

Comment:
I don't love this workaround but I can't think of a better option.

I used `exclude_defaults` to stop `model_dump` from serializing optional fields as explicit `None` values. We could use `exclude_none` instead, but I think apart from `apiVersion` and `kind` we really don't want to include defaults - since we want to leave the defaulting to server-side apply.

Reply:
people will lose the information which default values are available - in lsp - mhm

Reply:
apiVersion and kind are special - so i don't think the solution at all is a problem